data/pigui/modules/flight-ui/reticule.lua

```diff
@@ -420,23 +461,289 @@ local function displayManeuverData(radius)
+	-- info:
+	local flightstate = player:GetFlightState()
+	local flightcontrolstate = player:GetFlightControlState()
+	local followtarget = player:GetFollowTarget()
+	local navtarget = player:GetNavTarget()
+	local cmbtarget = player:GetCombatTarget()
+	local frmtarget = player.frameBody
+	local followmode = player:GetFollowMode()
+	local cruisedir = player:GetCruiseDirection()
+
+	local main_icon = nil -- must be overriden
+	if flightcontrolstate == "CONTROL_AUTOPILOT" then
+		local cmd = player:GetCurrentAICommand()
+		local ci = aicommand_info[cmd]
+		main_icon = icons.backward
+		tooltip = "UNKNOWN AI COMMAND?"
+		if ci then
+			main_icon = ci.icon
+			tooltip = ci.tooltip
+		end
+	elseif flightcontrolstate == "CONTROL_FIXSPEED" then
+		main_icon = cruisedir == "CRUISE_FWD" and icons.cruise_fwd or icons.cruise_up
+		tooltip = cruisedir == "CRUISE_FWD" and lui.HUD_BUTTON_CRUISE_FORWARD or lui.HUD_BUTTON_CRUISE_UP
+	else
+		main_icon = flightstate_info[flightcontrolstate].icon
+		tooltip = flightstate_info[flightcontrolstate].tooltip
+	end
+
+	-- background
+	ui.addIcon(pos, icons.follow_fill, back_color, Vector2(button_size, button_size), ui.anchor.center, ui.anchor.center, tooltip)
+	ui.addIcon(pos, icons.follow_edge, back_color, Vector2(button_size, button_size), ui.anchor.center, ui.anchor.center, tooltip)
+	-- edge icon
+	if followtarget then
+		local follow_icon = followmode == "FOLLOW_POS" and icons.follow_pos or icons.follow_ori
+		ui.addIcon(pos, follow_icon, colors.white, Vector2(button_size, button_size), ui.anchor.center, ui.anchor.center, tooltip)
+	end
+
+	-- main icon
+	ui.addIcon(pos, main_icon, colors.white, Vector2(icon_size, icon_size), ui.anchor.center, ui.anchor.center, tooltip)
+
+	-- radial menu
+	if ui.isMouseClicked(0) and ui.canClickOnScreenObjectHere() and (ui.getMousePos() - pos):length() < button_size / 2.0
+		or Input.IsActionActivated(bindings.assistRadial.action) then
+		local color_active = colors.white
+		local color_inactive = colors.lightBlackBackground
+		local color_disable = colors.white:opacity(0.3)
+
+		local icon_left = { icon = icons.backward, tooltip = "NO_ACTION", action = function(_) end, color = color_inactive }
+		local icon_right = { icon = icons.backward, tooltip = "NO_ACTION", action = function(_) end, color = color_inactive }
+		local icon_up = { icon = icons.backward, tooltip = "NO_ACTION", action = function(_) end, color = color_inactive }
+		local icon_down = { icon = icons.backward, tooltip = "NO_ACTION", action = function(_) end, color = color_inactive }
+
+		local MAX_FOLLOW_DISTANCE = 500000000 -- meters
+
+		-- up/down icons
+		if flightcontrolstate == "CONTROL_MANUAL" or flightcontrolstate == "CONTROL_FIXSPEED" then
+			icon_up.icon = icons.circ_cruise_fwd
+			icon_up.tooltip = lui.HUD_BUTTON_CRUISE_FORWARD
+			icon_up.color = color_active
+			icon_up.action = function(_)
+				player:SetCruiseDirection("CRUISE_FWD")
+				player:SetFlightControlState("CONTROL_FIXSPEED")
+			end
```

Comment:
Do all of these actions need to be generated inside the function? This seems like a mostly-invariant table that could take a reference to the player through the unused argument and be defined in a top-level table. Handling the different permutations of the radial menu could be done the same way, just using references to the different radial items when generating the radial table. It's not a major performance concern to have this code here, but it does feel a bit messy.

Reply:
Do I understand correctly, you propose that this function of 2 lines be taken out somewhere outside, and wrapped in some name?
And in this case, when the function is one line?
```
fix_icon.action = function(_) player:SetFlightControlState("CONTROL_FIXSPEED") end
```

Reply:
I'm mentioning more that the entire `icon_up` table be defined once at file scope and referenced in the function to build the table of buttons for the radial menu. As far as I can tell, there are a total of six radial menu buttons in use here, none of which actually have internals that need to be generated at runtime. Only the selection of which buttons to display changes between invocations of this radial menu.

Reply:
@Web-eWorks Ah, now I understand, thanks! What to do with dynamic tooltips?
https://github.com/Gliese852/pioneer/blob/838d73017c5852c1b2e9e13632955e746cd3b30c/data/pigui/modules/flight-ui/reticule.lua#L620

Reply:
Hmm... I overlooked that you were dynamically generating tooltips for two of the radial buttons. You *could* just change the tooltip field of the static table right before showing the radial menu, although that's not a particularly clean solution.

Probably the best solution would be to still move the invariant radial button tables to file scope and only generate the dynamic tooltip buttons in the function itself; that would still reduce runtime complexity but maintain flexibility. Alternatively you could avoid dynamic tooltips entirely, as the format of `Mars: Not Available` doesn't actually tell the player anything about what the button is supposed to do.

Reply:
@Web-eWorks I still think that specifying the object for the action is important, because the player can have many different types of targets (combat, navigation, autopilot, frame) and it would be nice to let him know which target he will follow. Therefore, I am for the first option - partially move it in a static table.
Perhaps I should also change the phrase "Not available" to "Follow orientation not available" and "Follow position not available" respectively.

Reply:
I moved something, I hope it got better. Only today I realized that this code is executed once, at the moment of clicking, so the difference in performance seems negligible.

Reply:
Yes, I'm not entirely concerned about the performance cost of the function (although every bit helps with immediate-mode UI), but more about the source-code complexity of the function; less code being executed at runtime means it's easier to debug and understand what's going on in the function. I'm happy with the updated function, thank you!